core/src/main/java/org/elasticsearch/plugins/ClusterPlugin.java

```diff
@@ -63,6 +64,12 @@
      * Called when the node is started
      */
     default void onNodeStarted() {
-
     }
+
+    /**
+     * Returns a map of {@link ClusterState.Custom} supplier that should be invoked to initialize the initial clusterstate.
+     * This allows custom clusterstate extensions to be always present and prevents invariants where clusterstates are published
+     * but customs are not initialized.
+     */
+    default Map<String, Supplier<ClusterState.Custom>> getInitialCustomSupplier() { return Collections.emptyMap(); }
```

Comment:
I think this plugin is the right place for it. I agree we should move others here.